server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -288,6 +288,15 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
         Property.NodeScope
     );
 
+    // value 0 can disable dynamic pruning optimization in cardinality aggregation
+    public static final Setting<Integer> CARDINALITY_AGGREGATION_PRUNING_THRESHOLD = Setting.intSetting(
+        "search.dynamic_pruning.cardinality_aggregation.max_allowed_cardinality",
+        100,
```

Comment:
let's run some benchmark to decide on the default here

Reply:
This is better to be a follow up task, separating from this one.
Understand the performance regression as cardinality goes up is a good experiment to do itself, and we can publish some guidance to public doc or blog later on.
Here I just want it to be a safe value, and user can update according to their will.